### ISSUE TITLE
Create Historical Volatility as an Indicator Class to be used on any strategy

### ISSUE THREAD

**Issue by phbrgnomo** [CoinAlpha/hummingbot#3629]
### What must be implemented

A historical volatility calculation function that can be used on any strategy.

It must use the similar frame of the volatility function implemented on the `avellaneda_market_making`, using the `BaseTrailingIndicator` class to populate the price buffer, and the use the buffer to calculate the historical volatility.

--------------
### Issue Details

Currently two strategies make use volatility calculation: `avellaneda_market_making` and `liquidity_mining`.

But both have the volatility calculated in a different way:

-----------
The `avellaneda_market_making` uses the `AverageVolatilityIndicator` class:

It execute the volatility calculation in percentage by:

```py
volatility_pct = self._avg_vol.current_value / float(self.get_price()) * 100.0
```

This calculation compares the difference between the quadratic mean of the price with the current price,

This can be seen as an "instant" volatility calculation. On short time frames (small buffer size) it's values is close to historical volatility with a faster "reaction" to faster price changes.

![Captura de tela 2021-06-24 12.21.31.png](https://images.zenhubusercontent.com/5f64254ca13ff3717cc3a0c5/9212f5b4-29be-46c4-98f2-a0f8ad31d155)
(Red line is historical volatility, blue line is "instant" volatility, with **10** samples)

But as mentioned on the [issue 2558](https://github.com/CoinAlpha/hummingbot/issues/2558), on longer timeframes (bigger buffer size), the "memory" of the past volatility values stay for longer, and the result moves too far from historical volatility.

![Captura de tela 2021-06-24 11.51.45.png](https://images.zenhubusercontent.com/5f64254ca13ff3717cc3a0c5/99a61c16-202b-4779-a8b6-c07513a03214)
(Red line is historical volatility, blue line is "instant" volatility, with **50** samples)

![Captura de tela 2021-06-24 11.59.06.png](https://images.zenhubusercontent.com/5f64254ca13ff3717cc3a0c5/1dcd5ad8-7d5d-4cfc-ad25-70166f4ad3a6)
(Red line is historical volatility, blue line is "instant" volatility, with **3600** samples)

-------------
On the `liquidity_mining` strategy, the volatility calculation is:

```py
            for i in range(last_index, first_index, self._volatility_interval * -1):
                prices = mid_prices[i - self._volatility_interval + 1: i + 1]
                if not prices:
                    break
                atr.append((max(prices) - min(prices)) / min(prices))
            if atr:
                self._volatility[market] = mean(atr)
```

This volatility calculated value is the average of the price range of the sample (`(max_price - min_price) / min)price`).

It is more similar to the calculation of an ATR (Average True Range) value than historical volatility.

------------------

Altough the above calculations works for their own pourposes, the correct historical volatility calculation (in percentage) is defined by:

![Captura de tela 2021-06-24 09.44.07.png](https://images.zenhubusercontent.com/5f64254ca13ff3717cc3a0c5/c90cd174-5781-4edf-9506-887b46593249)

It can also be calculated with:

```

price = [0, 1, ..., i-1, i]

hv = stdev( log(price[i] / price[i-1]) ) * 100

```

On the case of market prices, the volatility value returned is relative to the time frame of the data:
  - if the data is the price of every second, the calculation result is the "1 second volatility"
  - if the data is the price of every hour, the calculation result is the "hourly volatility"

This value can be transformed for different timeframes, by multiplying it by the square root of the wanted period
  - "1 second volatility" can be transformed into "1 minute volatility" by `sqrt(60) * calculated_volatility`
  - "hourly volatility" can be transformed into "daily volatility" by `sqrt(24) * calculated_volatility`

------------------

### Definition of done

**Based on the historical volatility formula, we must create a class similar to `AverageVolatilityIndicator` that:**

- Is named `HistoricalVolatilityIndicator`
- populate the price buffer with an array of price values
- apply the historical volatility calculation to the price buffer array
- returns the historical volatility value
- can have the price buffer updated as needed, by one of the following methods:
  - Add a new value to the price buffer every tick (like what `AverageVolatilityIndicator` already does)
  - repopulate the buffer from the start when a update is called (clean the buffer and add a new set of values)
- have a function that can be called to transform the volatility to different timeframes

**Also, for code functionality clarification the following is needed:**

- `AverageVolatilityIndicator` must be renamed to `InstantVolatilityIndicator`

---------------------------

### Followup

The following task will be created as a followup of the conclusion of the current one:

- Refactor the volatility used by `liquidity_mining` strategy as another Indicator class
- Use the Historical Volatility calculation to open new liquidity positions on the `uniswap_v3_lp` strategy

**Comment by keithbaum** [comment]
Create PR https://github.com/CoinAlpha/hummingbot/pull/3640 for this issue

**Comment by nicolas-powertrade** [comment]
Did not implement this point:
> * have a function that can be called to transform the volatility to different timeframes

As this is external to the volatility indicator and it relies on the user implementing the indicator to know how to interpret the result. Indicator know nothing about which timeframe are the samples coming from.

**Comment by phbrgnomo** [comment]
@vic-en for the volatility calculation lets use the implementation @keithbaum implemented on PR #3640.

While the results of both implementations should be the same, for code calculation use log different works better, resulting in less approximation errors.

Adjust PR #3646 as needed after #3640 get merged.

Other comments were added on the PR #3646.

about @nicolas-powertrade comment:

> Did not implement this point:
>> have a function that can be called to transform the volatility to different timeframes

This must be a function that should be called outside of the volatility function, to transform the return value.

For the uni_v3 for example, the price data is for every second, and to be useful for position creation (issue #3284), we must be able to transform the "second" volatility into an "annualized/monthly/daily" volatility.

The idea is to have a method/function that can be called on the volatility returned value to make it more practical to do the transformation:

For example:

```py 
calculated_volatility = #calculated volatility of the buffer sample data

calculated_volatility.change_timeframe(3600) # would transform the second volatility from uniswap prices to daily volatility, multiplying calculated_volatility by sqrt(3600)
```
